### ISSUE TITLE
Allow to pass other channel types

### ISSUE THREAD

**Issue by ddimtirov** [paritytrading/philadelphia#236]
In our testing, we would like to use pipe channels rather than socket channels. 

Looking at the uses of the channel, we should be able to swap the `SocketChannel` type in `FIXConnetion` with `ByteChannel`.

There would be negative impact of breaking the `SocketChannel FIXConnection::getChannel()` (it would change the return type)

Please indicate whether you would consider a Pull Request.

**Comment by jvirtanen** [comment]
Hi @ddimtirov! This is an interesting idea, and I could see other use cases for this as well. However, note that [`FIXConnection::send()`](https://github.com/paritytrading/philadelphia/blob/2052dd027ef65d9f4fec37629ab09e2d88c6612e/libraries/core/src/main/java/com/paritytrading/philadelphia/FIXConnection.java#L452) uses [`ScatteringByteChannel`](https://docs.oracle.com/javase/8/docs/api/java/nio/channels/ScatteringByteChannel.html) as well.

What do you think if we made `FIXConnection` generic with a type parameter for the concrete channel implementation?
```java
public class FIXConnection<C extends ByteChannel & ScatteringByteChannel> {
    ...
```

One downside I see is that one would need to start using the diamond syntax for a `FIXConnection` instance:
```java
FIXConnection<SocketChannel> connection = new FIXConnection<>(...);  
```

On the other hand, this would unlock pipe channels and, interestingly, even [`DatagramChannel`](https://docs.oracle.com/javase/8/docs/api/java/nio/channels/DatagramChannel.html). 😅

**Comment by ddimtirov** [comment]
Sounds good, would you like me to try it?

**Comment by jvirtanen** [comment]
Please go ahead, let's see how it would look like in practice. 🙂

**Comment by mrossi975** [comment]
hi guys,
the above would also allow plugging in the TLS channel implemented by the following open-source project:
https://github.com/marianobarrios/tls-channel
This is important in order to connect to CoinBase. 
Having a separate TLS tunnel in a separate tunnel is OK as a proof-of-concept but it's not great for a production environment, you have to worry about configuring, deploying and running one more process.
I am happy to help with all of the above guys..

**Comment by ddimtirov** [comment]
Update - the patch is in the mail (let's see how long our internal processes would take).

**Comment by mrossi975** [comment]
hey guys, any update on the above? I'd be keen to try this out. Happy to do any work required. Thanks!

**Comment by jvirtanen** [comment]
This was implemented in #238.
